### PR TITLE
Cf 9266 - Provide batch sync service to handle large volumes of data

### PR DESCRIPTION
**Pull request by phuminhphan**
https://procoretech.atlassian.net/browse/CF-9266

Batching support.  Very necessary for ERP as there will be instances where we are syncing over 2000+ vendors at one time to Procore.  Currently, we timeout or get random `400 Bad request` errors in our Quickbooks integration.

### CODE REVIEW

File: lib/procore/services/batch_sync.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+        { entities: entities, errors: errors }
+      end
+
+      private
+
+      attr_reader :url, :arguments, :updates, :connection, :batch_size
+
+      def batches
+        updates.in_groups_of(batch_size)
```
Comment (AllPurposeName) [inline lib/procore/services/batch_sync.rb]: I believe we want to add an argument of `false` so as not to send a large number of unnecessary `nil`s. https://apidock.com/rails/Array/in_groups_of

```suggestion
        updates.in_groups_of(batch_size, false)
```

File: test/procore/services/batch_sync_test.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Procore::Services::BatchSyncTest < Minitest::Test
+  def test_execute
+    updates = (0..1000).to_a.map { |count| { id: count, origin_id: "origin_id_#{count}" } }
+    oauth_connection = Minitest::Mock.new
+
+    url = '/erp_projects/sync'
```
Comment (AllPurposeName) [inline test/procore/services/batch_sync_test.rb]: Even though it's just an example, let's find a public version of the sync endpoint to use here.

File: test/procore/services/batch_sync_test.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      ]
+    }
+    assert_equal(sync_result, expected_sync_result)
+  end
+
+  def test_execute_with_custom_batch_size
+    updates = (0..5).to_a.map { |count| { id: count, origin_id: "origin_id_#{count}" } }
+    oauth_connection = Minitest::Mock.new
+
+    url = '/erp_projects/sync'
```
Comment (AllPurposeName) [inline test/procore/services/batch_sync_test.rb]: Same as above.

File: lib/procore/services/batch_sync.rb (hunk 1 above)
Comment (AllPurposeName) [reply lib/procore/services/batch_sync.rb]: Alternatively we could just test with `nil`s in the updates.

File: lib/procore/services/batch_sync.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        errors = []
+        batches.each do |batch|
+          sync_arguments = arguments.merge(updates: batch.compact)
+          response = connection.patch(url, sync_arguments).body
+          entities += response['entities'] if response['entities']&.is_a?(Array)
+          errors += response['errors'] if response['errors']&.is_a?(Array)
+        end
+
+        { entities: entities, errors: errors }
+      end
```
Comment (AllPurposeName) [inline lib/procore/services/batch_sync.rb]: NB: I think we could also change this block to a better enumerable like a reduce or each_with_object.

I'm thinking something like:

```
def execute
  batches.each_with_object({entities:[],errors:[]}) do |batch, results|
    sync_arguments = arguments.merge(updates: batch.compact)
    response = connection.patch(url, sync_arguments).body
    results[:entities] += response['entities'] if response['entities']&.is_a?(Array)
    results[:errors] += response['errors'] if response['errors']&.is_a?(Array)
  end
end
```

What do you think?

Comment (ylansegal) [reply lib/procore/services/batch_sync.rb]: I like @AllPurposeName idea of using `each_with_object`

**Comment by phuminhphan** [comment]
@PatKoperwas  - `arguments` is basically the `options` that you're mentioning above. I will rename it.

I wouldn't want `batch_sync` in the `Client` as the client is a set of standard `patch` `put` `post` etc... methods.  `batch_sync` is a `patch`.

I was thinking of this as more of a `Service` type of class where it helps the consumer do one specific thing.

**Comment by phuminhphan** [comment]
@PatKoperwas if the point of this gem is to be primitive calls, then I think i'll just close this PR and put this capability into our own gem we developed for ERP. 
Thanks for the review!

**Pull request closed without merging**